### PR TITLE
feat(general): add severity metadata to custom policy

### LINKED ISSUE #884: Add Severity to controls and option to specify the severity to be reported.
**Is your feature request related to a problem? Please describe.**
Currently, checkov only report success for failure. That would be great if you can add severity to each control
also add options to specify the desired severity level of violations to be reported.

**Additional context**
You can get severity of CIS rules for aws at https://docs.aws.amazon.com/securityhub/latest/userguide/securityhub-cis-controls.html
Implementation in terrascan https://github.com/accurics/terrascan/pull/501

### PR DESCRIPTION
**Pull request by dtrouillet**
**By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

[//]: # "
    # PR Title
    We use the title to create changelog automatically and therefore only allow specific prefixes
    - break:    to indicate a breaking change, this supersedes any of the other types
    - feat:     to indicate new features or checks
    - fix:      to indicate a bugfix or handling of edge cases of existing checks
    - docs:     to indicate an update to our documentation
    - chore:    to indicate adjustments to workflow files or dependency updates
    - platform: to indicate a change needed for the platform
    Each prefix should be accompanied by a scope that specifies the targeted framework. If uncertain, use 'general'.
    #    
    Allowed prefixs:
    ansible|argo|arm|azure|bicep|bitbucket|circleci|cloudformation|dockerfile|github|gha|gitlab|helm|kubernetes|kustomize|openapi|sast|sca|secrets|serverless|terraform|general|graph|terraform_plan|terraform_json
    #
    ex.
    feat(terraform): add CKV_AWS_123 to ensure that VPC Endpoint Service is configured for Manual Acceptance
"

## Description

*For custom policies, we need to add information about severity. This severity can be other than Unknown in offline (airgapped environment)*

I just add metadata.severity object in yaml custom policy definition.

Fixes #884

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my feature, policy, or fix is effective and works
- [x] New and existing tests pass locally with my changes

### CODE REVIEW

**Comment by aamchi78** [comment]
Great, this exactly what i need.

**Comment by itai1357** [comment]
Hi @dtrouillet!
Thanks a lot for contributing 

the following unit-tests are failing 
```
tests/arm/checks/resource/test_SkipJsonRegexPattern.py::TestSkipJsonRegexPattern::test_no_skip tests/arm/checks/resource/test_SkipJsonRegexPattern.py::TestSkipJsonRegexPattern::test_skip_specific_check
tests/arm/checks/resource/test_SkipJsonRegexPattern.py::TestSkipJsonRegexPattern::test_skip_specific_check_for_folder 
tests/arm/checks/resource/test_SkipJsonRegexPattern.py::TestSkipJsonRegexPattern::test_skip_specific_check_specific_file
```

seems that somehow even passed-checks become failed with this change.
we will appreciate your awareness for it (-:

**Comment by dtrouillet** [comment]
Hi @itai1357 ,

Theses unit tests are in failure on main branch without my pr : https://github.com/bridgecrewio/checkov/actions/runs/9988292968/job/27604459450 

image

**Comment by itai1357** [comment]
Hi @dtrouillet.
TNX you are right. my bad

**Comment by dtrouillet** [comment]
@itai1357 , @achiar99 , can you review this PR please?

**Comment by dtrouillet** [comment]
@Saarett or @ChanochShayner  can you review this pr please.

File: docs/3.Custom Policies/YAML Custom Policies.md [hunk 1]
```diff
@@ -21,6 +21,7 @@ The Metadata includes:
 * ID - `CKV2_<provider>_<number>`
 * Category
 * Guideline (optional)
+* Severity (optional) - can be `NONE`, `INFO`, `LOW`, `MEDIUM`, `MODERATE`, `HIGH`, `IMPORTANT`, `CRITICAL`, `OFF`
```
Comment (ChanochShayner) [inline docs/3.Custom Policies/YAML Custom Policies.md:24]: The severities should be - 
`INFO`, `LOW`, `MEDIUM`, `HIGH`, `CRITICAL`

**Review by ChanochShayner — COMMENTED** [review]
Looks good.
Please see my comment, and please create a real UT to ensure the severity is in the custom policy after the scan.

**Comment by dtrouillet** [comment]
@ChanochShayner , I just added UT and fix documentation

File: docs/3.Custom Policies/YAML Custom Policies.md [hunk 2]
```diff
@@ -35,14 +36,27 @@ The possible values for category are:
 * KUBERNETES
 * APPLICATION_SECURITY
 * SUPPLY_CHAIN
-* API_SECURITY
+* API_SECURITY 
+
+The possible values for severity are:
+
+* NONE
```
Comment (ChanochShayner) [inline docs/3.Custom Policies/YAML Custom Policies.md:43]: Please adjust according to previous list

File: checkov/common/checks_infra/checks_parser.py [hunk 3]
```diff
@@ -201,6 +202,7 @@ def parse_raw_check(self, raw_check: Dict[str, Dict[str, Any]], **kwargs: Any) -
         check.name = raw_check.get("metadata", {}).get("name", "")
         check.category = raw_check.get("metadata", {}).get("category", "")
         check.frameworks = raw_check.get("metadata", {}).get("frameworks", [])
+        check.severity = get_severity(raw_check.get("metadata", {}).get("severity", ""))
```
Comment (ChanochShayner) [inline checkov/common/checks_infra/checks_parser.py:205]: Could you please check first if the severity exists?
It is better to have this field just if we have a value for it, and not always empty string.

File: docs/3.Custom Policies/YAML Custom Policies.md (hunk 2 above)
Comment (dtrouillet) [reply docs/3.Custom Policies/YAML Custom Policies.md]: Oups, Sorry, it's fixed

File: checkov/common/checks_infra/checks_parser.py (hunk 3 above)
Comment (dtrouillet) [reply checkov/common/checks_infra/checks_parser.py]: Ok. I just added a condition

**Comment by dtrouillet** [comment]
@ChanochShayner I did the modifications

File: checkov/common/checks_infra/checks_parser.py [hunk 4]
```diff
@@ -201,6 +202,9 @@ def parse_raw_check(self, raw_check: Dict[str, Dict[str, Any]], **kwargs: Any) -
         check.name = raw_check.get("metadata", {}).get("name", "")
         check.category = raw_check.get("metadata", {}).get("category", "")
         check.frameworks = raw_check.get("metadata", {}).get("frameworks", [])
+        severity = get_severity(raw_check.get("metadata", {}).get("severity", ""))
```
Comment (ChanochShayner) [inline checkov/common/checks_infra/checks_parser.py:205]: Sould be like - 
```
        if raw_check.get("metadata", {}).get("severity"):
            severity = get_severity(raw_check.get("metadata", {}).get("severity"))
```

File: checkov/common/checks_infra/checks_parser.py [hunk 5]
```diff
@@ -201,6 +202,9 @@ def parse_raw_check(self, raw_check: Dict[str, Dict[str, Any]], **kwargs: Any) -
...
+        if severity != "":
```
Comment (ChanochShayner) [inline checkov/common/checks_infra/checks_parser.py:206]: Please fix the mypy errors on this line.

**Comment by dtrouillet** [comment]
@ChanochShayner , I fixed the mypy issue and I added an UT for empty severity in custom policy

File: checkov/common/checks_infra/checks_parser.py [hunk 6]
```diff
@@ -201,6 +202,9 @@ def parse_raw_check(self, raw_check: Dict[str, Dict[str, Any]], **kwargs: Any) -
         check.name = raw_check.get("metadata", {}).get("name", "")
         check.category = raw_check.get("metadata", {}).get("category", "")
         check.frameworks = raw_check.get("metadata", {}).get("frameworks", [])
+        severity = get_severity(raw_check.get("metadata", {}).get("severity", ""))
+        if severity != None:
```
Comment (ChanochShayner) [inline checkov/common/checks_infra/checks_parser.py:206]:
```suggestion
        if severity:
```

File: tests/test_main.py [hunk 7]
```diff
@@ -143,3 +143,38 @@ def test_run_with_severity_skip_filter_without_api_key(caplog: LogCaptureFixture
...
+    argv = ["-d", str(resource_dir), "--framework", "terraform", "--external-checks-dir", str(extra_checks_dir_path), "--check", "CUSTOM_SEVERITY"]
+
+    # when
+    ckv = Checkov()
+    ckv.parse_config(argv=argv)
+    ckv.run(banner=custom_banner)
+
+    # then
+    for report in ckv.scan_reports:
+        print(report.failed_checks[0].check_id)
```
Comment (ChanochShayner) [inline tests/test_main.py:160]:
```suggestion please remove
```

File: tests/test_main.py [hunk 8]
```diff
@@ -143,3 +143,38 @@ def test_run_with_severity_skip_filter_without_api_key(caplog: LogCaptureFixture
...
+    argv = ["-d", str(resource_dir), "--framework", "terraform", "--external-checks-dir", str(extra_checks_dir_path), "--check", "CUSTOM_WITHOUT_SEVERITY"]
+
+    # when
+    ckv = Checkov()
+    ckv.parse_config(argv=argv)
+    ckv.run(banner=custom_banner)
+
+    # then
+    for report in ckv.scan_reports:
+        print(report.failed_checks[0].check_id)
```
Comment (ChanochShayner) [inline tests/test_main.py:178]:
```suggestion
```

**Comment by dtrouillet** [comment]
@ChanochShayner  sorry for the many mistakes, Python is clearly not my favorite language :/

**Review by ChanochShayner — APPROVED** [review]
Looks good!

**Comment by dtrouillet** [comment]
@itai1357, @omryMen , @achiar99 , can you review this PR please?

**Review by matansha — APPROVED** [review]

**Pull request merged**